apps/labs/posts/numpy-python-api-cleanup.md

```diff
@@ -0,0 +1,345 @@
+referring to, and we make an explicit declaration about precision, leaving no room for
+platform-specific behavior. "Words" that refer to C types should be used when
+interacting with C exctension modules directly.
+
+### Ruff plugin
+
+Ruff is a new Python linter that can outperform other well-known linting tools.
+During one of the discussions on Scientific Python Discord server, we came up with
+the idea of writing a dedicated Ruff rule to address NumPy 2.0 changes. The rule offers
+an automated way to fix a large part of changes - migrating to retained aliases or
+flagging lines that require manual intervention to be compatible with NumPy 2.0.
+
+This assignment gave me the opportunity to write a bit of Rust code, since it is
+the core language for Ruff. As of today, the PR is still open, but the rule is available
+in the local setup and it managed to correctly fix over a hundred of lines of code for
+the latest SciPy release source code.
+
+The command for running Ruff linter:
+
+```sh
+ruff check scipy/ --no-cache --fix
+```
+
+And here's a small passage from `git diff`:
+
+```diff
+-     	a = toarray(a, dtype=np.float_)
++     	a = toarray(a, dtype=np.float64)
+...
+-	elif (b == Inf and a == -Inf):
++	elif (b == np.inf and a == -np.inf):
+...
+-	math_dtypes = [np.int_, np.float_, np.complex_]
++	math_dtypes = [np.int_, np.float64, np.complex128]
+```
+
+Ruff rules open more robust ways to apply systematic changes to large codebases,
+compared to regular expression searches, because they offer AST analysis rather
+than a text-based search only.
+
+---
+
+## Wrapping up
+
+The development and maintenance efforts for NumPy package will continue, most notably
+in the Array API area, where full compatibility is still being implemented. The release
+of NumPy 2.0 is planned for end of the year, and only then will the official adoption
+of the new, major version begin.
+
+---
+
+## Acknowledgements
+
+I would like to thank my mentors [Ralf Gommers](https://github.com/rgommers) and
+[Nathan Goldbaum](https://github.com/ngoldbaum) for their advices and guidance during
+the whole internship, and [Melissa Weber Mendonça](https://github.com/melissawm) for
+organising and conducting intern cohort meetings. The time spent on NEP 52 was a perfect
+primer to the Scientific Python ecosystem!
+
+I look forward to continuing working on NumPy and other libraries within the community!
+
+---
+
+## References
+
+- [NEP 52](https://numpy.org/neps/nep-0052-python-api-cleanup.html)
+- [NumPy 2.0 migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html)
+- [Main namespace cleanup tracking issue](https://github.com/numpy/numpy/issues/24306)
+- [`numpy.lib` cleanup tracking issue](https://github.com/numpy/numpy/issues/24507)
+- [`core` to `_core` refactor PR](https://github.com/numpy/numpy/pull/24634)
+- [Ruff NumPy 2.0 rule PR](https://github.com/astral-sh/ruff/pull/7702)
+
+<br />
```

Comment:
should this html thingy be here?

Reply:
I think I found it in some template file in the repo but I think it's redundant - removed.